pymc3/distributions/multivariate.py

```diff
@@ -2089,3 +2090,119 @@ def logp(self, value):
+    Mean      :math:`\mu`
+    Variance  :math:`(\tau D(I-\alpha W))^{-1}`
+    ========  ==========================
+
+    Parameters
+    ----------
+    mu: array
+        Mean vector
+    W: matrix
+        Symmetric adjacency matrix of 1s and 0s indicating
+        adjacency between elements.
+    alpha: float or array
+        Autoregression parameter. Values closer to 0 indicate weaker
+        correlation and values closer to 1 indicate higher autocorrelation.
+    tau: float or array
+        Precision variables controlling the scale of the underlying normal variates.
+    sparse: bool, default=False
+        Determines whether or not sparse computations are used
+
+    References
+    ----------
+    ..  Jin, X., Carlin, B., Banerjee, S.
+        "Generalized Hierarchical Multivariate CAR Models for Areal Data"
+        Biometrics, Vol. 61, No. 4 (Dec., 2005), pp. 950-961
+    """
+
+    def __init__(self, mu, W, alpha, tau, sparse=False, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        D = W.sum(axis=0)
+        d, _ = W.shape
+
+        self.d = d
+        self.median = self.mode = self.mean = self.mu = aet.as_tensor_variable(mu)
+        self.sparse = sparse
+
+        if not W.ndim == 2 or not np.allclose(W, W.T):
+            raise ValueError("W must be a symmetric adjacency matrix.")
+
+        if sparse:
+            W_sparse = scipy.sparse.csr_matrix(W)
+            self.W = aesara.sparse.as_sparse_variable(W_sparse)
+        else:
+            self.W = aet.as_tensor_variable(W)
+
+        # eigenvalues of D^−1/2 * W * D^−1/2
+        Dinv_sqrt = np.diag(1 / np.sqrt(D))
+        DWD = np.matmul(np.matmul(Dinv_sqrt, W), Dinv_sqrt)
+        self.lam = scipy.linalg.eigvalsh(DWD)
+        self.D = aet.as_tensor_variable(D)
+
+        tau = aet.as_tensor_variable(tau)
+        if tau.ndim > 0:
+            self.tau = tau[:, None]
+        else:
+            self.tau = tau
+
+        alpha = aet.as_tensor_variable(alpha)
+        if alpha.ndim > 0:
+            self.alpha = alpha[:, None]
+        else:
+            self.alpha = alpha
+
+    def logp(self, value):
+        """
+        Calculate log-probability of a CAR-distributed vector
+        at specified value.
+
+        Parameters
+        ----------
+        value: numeric
+            Value for which log-probability is calculated.
+
+        Returns
+        -------
+        TensorVariable
+        """
+
+        if value.ndim == 1:
+            value = value[None, :]
+
+        logtau = self.d * aet.log(self.tau).sum()
+        logdet = aet.log(1 - self.alpha.T * self.lam[:, None]).sum()
+        delta = value - self.mu
+
+        if self.sparse:
+            Wdelta = aesara.sparse.dot(delta, self.W)
+        else:
+            Wdelta = aet.dot(delta, self.W)
+
+        tau_dot_delta = self.D[None, :] * delta - self.alpha * Wdelta
+        logquad = (self.tau * delta * tau_dot_delta).sum(axis=-1)
+        return 0.5 * (logtau + logdet - logquad)
```

Comment:
Do you need bounds on the logp for invalid parameter values? (E.g. negative tau) or does your logp naturally return -inf for that?

Reply:
I didn't realize that many of the distributions have their `logp` functions passed through `bound`. I'm going to have to include that as well - this function does nothing to handle invalid parameter values.